### DISCUSSION TITLE
Shell Spell: Extracting and Propagating Multiple Values With jq - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#318] [Announcements]
# Shell Spell: Extracting and Propagating Multiple Values With jq - Gunnar Morling

In my day job at Decodable,
I am currently working with Terraform to provision some cloud infrastructure for an upcoming hands-on lab.
Part of this set-up is a Postgres database on Amazon RDS,
which I am creating using the Terraform AWS modules.
Now, once my database was up and running,
I wanted to extract two dynamically generated values from Terraform:
the random password created for the root user, and the database host URL.
On my way down the rabbit hole for finding a CLI command for doing this efficiently,
I learned a few interesting shell details which I’d like to share.

https://www.morling.dev/blog/extracting-and-propagating-multiple-values-with-jq/

**Comment by skagedal** [comment]
Nice solution, I wouldn't have thought of that!

One option which I think meets your requirements would be to store the `terraform` output in a shell variable. I came up with this which seems to work:

```
(resources="$(terraform show -json | jq '
    .values.root_module.child_modules[]
    | select(.address=="module.lab-001")
    | .resources[]
    ')"
host=$(echo $resources | jq -r 'select(.address=="module.lab-001.random_password.this") | .values.result')
password=$(echo $resources | jq -r 'select(.address=="module.lab-001.aws_db_instance.lab_001_db") | .values.address')
psql "postgresql://root:${password}@${host}:5432/labdb")
```

I think this option might have an advantage in debugability. Of course, doing it all as a single `jq` pipeline has a certain elegance about it.
